### PR TITLE
OHRM5X-554: Project Activity API Added

### PR DESCRIPTION
**Pull request by VageeshOrangeHRM**
This PR Contains following changes:
1. Project Activity API, Service and DAO created.
2. Entity created. (Customer, Project, Project Activity, Timesheet Item, Timesheet).
3. Unit test added for API, DAO and Service layers.
4. Data group permission added.
5. OHRM5X-985 - Description field related issues in add/edit customers form.

### CODE REVIEW

File: symfony/plugins/orangehrmTimePlugin/Api/Model/ProjectActivityModel.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+use OrangeHRM\Core\Api\V2\Serializer\ModelTrait;
+use OrangeHRM\Core\Api\V2\Serializer\Normalizable;
+use OrangeHRM\Entity\ProjectActivity;
+
+class ProjectActivityModel implements Normalizable
+{
+    use ModelTrait;
+
+    public function __construct(ProjectActivity $projectActivity)
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmTimePlugin/Api/Model/ProjectActivityModel.php:30]: missing return

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        $count = $this->getPaginator($q)->count();
+        return ($count > 0);
+    }
+
+    /**
+     * @param int[] $toBeDeletedActivityIds
+     * @throws ProjectServiceException
+     */
+    public function deleteProjectActivities(array $toBeDeletedActivityIds)
+    {
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:117]: missing return

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        $delete = true;
+        foreach ($toBeDeletedActivityIds as $toBeDeletedActivityId) {
+            $deletable = $this->hasActivityGotTimesheetItems($toBeDeletedActivityId);
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:120]: you can simplify this 
```
if(! $this->hasActivityGotTimesheetItems($toBeDeletedActivityId)) {
  throw ProjectServiceException::projectActivityExist();
}
```

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        return ($count > 0);
+    }
+
+    /**
+     * @param int[] $toBeDeletedActivityIds
+     * @throws ProjectServiceException
+     */
+    public function deleteProjectActivities(array $toBeDeletedActivityIds)
+    {
+        $delete = true;
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:118]: you don't need this

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        foreach ($toBeDeletedActivityIds as $toBeDeletedActivityId) {
+            $deletable = $this->hasActivityGotTimesheetItems($toBeDeletedActivityId);
+            if ($deletable) {
+                $delete = false;
+                break;
+            }
+        }
+        if ($delete) {
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:126]: this too

**Review by samanthajayasinghe — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php (hunk 3 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:120]: foreach ($toBeDeletedActivityIds as $toBeDeletedActivityId) {
            if ($this->hasActivityGotTimesheetItems($toBeDeletedActivityId)) {
                throw ProjectServiceException::projectActivityExist();
            }
        }

Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:120]: image

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php (hunk 2 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:117]: done

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php (hunk 4 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:118]: done

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php (hunk 5 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:126]: done

File: symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function update(): EndpointResult
+    {
+        $projectActivityId = $this->getRequestParams()->getInt(
+            RequestParams::PARAM_TYPE_ATTRIBUTE,
+            CommonParams::PARAMETER_ID
+        );
+        $projectId = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, self::PARAMETER_PROJECT_ID);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php:214]: can refactor
ref: `symfony/plugins/orangehrmPimPlugin/Api/EmployeeTerminationAPI.php` line no 53

File: symfony/plugins/orangehrmTimePlugin/Service/ProjectActivityService.php [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * @return ProjectActivityDao
+     */
+    public function getProjectActivityDao(): ProjectActivityDao
+    {
+        if (!$this->projectActivityDao instanceof ProjectActivityDao) {
+            $this->projectActivityDao = new ProjectActivityDao();
+        }
+        return $this->projectActivityDao;
+    }
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Service/ProjectActivityService.php:37]: Shall we move this to `ProjectService`

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml [hunk 8]
```diff
@@ -20,6 +20,20 @@ apiv2_time_validate_customer_name:
     _key: customerName
     _api: OrangeHRM\Time\Api\ValidationCustomerNameAPI
 
+apiv2_time_project_activities:
+  path: /api/v2/time/project/{projectId}/project-activity
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/config/routes.yaml:24]:
```suggestion
  path: /api/v2/time/project/{projectId}/activities
```

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml [hunk 9]
```diff
@@ -20,6 +20,20 @@ apiv2_time_validate_customer_name:
...
+  controller: OrangeHRM\Core\Controller\Rest\V2\GenericRestController::handle
+  methods: [ GET, POST, DELETE ]
+  defaults:
+    _api: OrangeHRM\Time\Api\ProjectActivityAPI
+
+apiv2_time_project_activity:
+  path: /api/v2/time/project/{projectId}/project-activity/{id}
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/config/routes.yaml:31]:
```suggestion
  path: /api/v2/time/project/{projectId}/activities/{id}
```

File: symfony/plugins/orangehrmTimePlugin/entity/Customer.php [hunk 10]
```diff
@@ -22,6 +22,8 @@
 use Doctrine\ORM\Mapping as ORM;
 
 /**
+ * Customer
+ *
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/entity/Customer.php:26]: no need

File: symfony/plugins/orangehrmTimePlugin/entity/Customer.php [hunk 11]
```diff
@@ -58,66 +60,73 @@ class Customer
     private string $description;
 
     /**
-     * @return string
+     * @var Project
+     *
+     * @ORM\OneToOne (targetEntity="OrangeHRM\Entity\Project", mappedBy="customer")
      */
-    public function getName(): string
+    private Project $project;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/entity/Customer.php:67]: is this mapping correct? I think customer can have many projects

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+use OrangeHRM\Time\Exception\ProjectServiceException;
+
+class ProjectActivityDao extends BaseDao
+{
+    /**
+     * @param int $projectId
+     * @param ProjectActivitySearchFilterParams $projectActivitySearchFilterParams
+     * @return Paginator
+     */
+    public function getProjectActivitiesByProjectId(
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:37]:
```suggestion
    private function getProjectActivitiesPaginator(
```

File: symfony/plugins/orangehrmTimePlugin/entity/ProjectActivity.php [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+     * @ORM\Column(name="name", type="string", length=110)
+     */
+    private string $name;
+
+    /**
+     * @var TimeSheetItem
+     *
+     * @ORM\OneToOne(targetEntity="OrangeHRM\Entity\TimeSheetItem", mappedBy="projectActivity")
+     */
+    private TimeSheetItem $timeSheetItem;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/entity/ProjectActivity.php:68]: seems mapping wrong

File: symfony/plugins/orangehrmTimePlugin/entity/TimeSheet.php [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+use DateTime;
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * TimeSheet
+ *
+ * @ORM\Table(name="ohrm_timesheet")
+ * @ORM\Entity
+ */
+class TimeSheet
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/entity/TimeSheet.php:31]:
```suggestion
class Timesheet
```

File: symfony/plugins/orangehrmTimePlugin/entity/TimeSheetItem.php [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+use DateTime;
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * TimeSheetItem
+ *
+ * @ORM\Table(name="ohrm_timesheet_item")
+ * @ORM\Entity
+ */
+class TimeSheetItem
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/entity/TimeSheetItem.php:31]:
```suggestion
class TimesheetItem
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+     */
+    public function create(): EndpointResult
+    {
+        $projectActivity = new ProjectActivity();
+        $this->setParamsToProjectActivity($projectActivity);
+        $this->getProjectActivityService()->getProjectActivityDao()->saveProjectActivity($projectActivity);
+        return new EndpointResourceResult(ProjectActivityModel::class, $projectActivity,);
+    }
+
+    public function setParamsToProjectActivity(ProjectActivity $projectActivity): void
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+    {
+        $projectId = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, self::PARAMETER_PROJECT_ID);
+        $activityName = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_NAME);
+        $project = $this->getProjectActivityService()->getProjectActivityDao()->getProjectById($projectId);
+        $projectActivity->setName($activityName);
+        $projectActivity->setProject($project);
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+     */
+    public function getOne(): EndpointResult
+    {
+        $projectActivityId = $this->getRequestParams()->getInt(
+            RequestParams::PARAM_TYPE_ATTRIBUTE,
+            CommonParams::PARAMETER_ID
+        );
+        $projectId = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, self::PARAMETER_PROJECT_ID);
+        $projectActivity = $this->getProjectActivityService()->getProjectActivityDao(
+        )->getProjectActivityByProjectIdAndProjectActivityId($projectId, $projectActivityId);
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php (hunk 6 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/ProjectActivityAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml (hunk 8 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/config/routes.yaml]: done

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml (hunk 9 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/config/routes.yaml]: done

File: symfony/plugins/orangehrmTimePlugin/entity/Customer.php (hunk 10 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/entity/Customer.php]: done

File: symfony/plugins/orangehrmTimePlugin/entity/TimeSheet.php (hunk 14 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/entity/TimeSheet.php]: done

File: symfony/plugins/orangehrmTimePlugin/entity/TimeSheetItem.php (hunk 15 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/entity/TimeSheetItem.php]: done

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**